### PR TITLE
feat(s3): constructs factories - create well architected s3 buckets

### PR DESCRIPTION
**Pull request by biffgaut**
*Description of changes:*
This PR introduces a brand new functionality in Solutions Constructs - Constructs Factories. All Solutions Constructs are built upon our underlying core library which creates well architected resources from a single service (which our L3 Solutions Constructs then connect). Constructs Factories expose our core library functionality in a documented, supported way and is the preferred way to access the functionality in our core library.

This first release creates a factory for S3 buckets, allowing creation of a bucket following all best practices with a single call. We plan to add SQS queues and Step Functions state machines in the near future.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 71347789190a201da9148b83a6282058265be3db
* Result: FAILED
* [Build Logs](https://t2ulkoluaf.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6f5ad316-5127-4c0b-bc86-9ac65d092622%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: d5dd0a2df645227e61a90357117041ee62aac1cd
* Result: FAILED
* [Build Logs](https://t2ulkoluaf.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b30044e1-7583-480d-8f32-b9c2f50fc8ba%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 69fdc8f907422d94e31a27b4b6c5b68bf6c17eca
* Result: FAILED
* [Build Logs](https://t2ulkoluaf.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2d943661-d4c7-47fa-8fed-ecf4cce2edd3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 6309dd17875c6c541bae13cc261bef5eb66d6a0b
* Result: SUCCEEDED
* [Build Logs](https://t2ulkoluaf.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d8efef22-df43-422f-adef-a644761a9f44%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/aws-constructs-factories/lib/index.ts
Comment (georgebearden) [inline source/patterns/@aws-solutions-constructs/aws-constructs-factories/lib/index.ts:29]: This is just a house keeping thing, but over time this list of inputs/outputs to the factories will get very large. Would it make more sense to move them to their own file, one file for each set? Just something to contemplate, not even sure if its the right answer.

File: source/patterns/@aws-solutions-constructs/aws-constructs-factories/lib/index.ts
Comment (georgebearden) [inline source/patterns/@aws-solutions-constructs/aws-constructs-factories/lib/index.ts:39]: It would be nice if `defaults.CheckS3Props` could handle undefined, to avoid having to put these calls in the factory methods.

File: source/patterns/@aws-solutions-constructs/aws-constructs-factories/lib/index.ts
Comment (georgebearden) [inline source/patterns/@aws-solutions-constructs/aws-constructs-factories/lib/index.ts:44]: as these fields are exactly the same, would we want to use the spread operator instead to simplify the logic?

```
const propsArg: defaults.BuildS3BucketProps = props ? { ...props } : {}
```

File: source/patterns/@aws-solutions-constructs/aws-constructs-factories/package.json
Comment (georgebearden) [inline source/patterns/@aws-solutions-constructs/aws-constructs-factories/package.json:10]: this directory looks wrong "aws-dynamodbstreams-lambda"

File: source/patterns/@aws-solutions-constructs/core/lib/s3-bucket-helper.ts
Comment (georgebearden) [inline source/patterns/@aws-solutions-constructs/core/lib/s3-bucket-helper.ts:176]: nice comment for redirection

**Review by georgebearden — DISMISSED** [review]
Few comments to think about, but overall - lgtm.

File: source/patterns/@aws-solutions-constructs/aws-constructs-factories/lib/index.ts
Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/aws-constructs-factories/lib/index.ts:44]: Yes, they are the same - but is that an fundamental trait or a coincidence? The factory is supposed to put an abstraction over the core function - feels like using the spread operator because of the coincidence will increase coupling between the exposed factory interface and the core function.

File: source/patterns/@aws-solutions-constructs/aws-constructs-factories/lib/index.ts
Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/aws-constructs-factories/lib/index.ts:29]: I thought about that - this approach will become absolutely unmanageable as we expose more factories and must change. But I postponed devising an approach to handle multiple factories until I had more examples (rather than trying to come up with the generic from a single example).

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: e00f9156984b83e7690a468c7589a19449d0aaca
* Result: SUCCEEDED
* [Build Logs](https://t2ulkoluaf.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=531f9e6c-0abf-4c9e-a500-0b624c5ec3a1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/aws-constructs-factories/package.json
Comment (biffgaut) [reply source/patterns/@aws-solutions-constructs/aws-constructs-factories/package.json]: Addressed

File: source/patterns/@aws-solutions-constructs/core/README.md
Comment (hayesry) [inline source/patterns/@aws-solutions-constructs/core/README.md:21]: super nit: two spaces between "Major" and "release" in "Major  release"

**Review by hayesry — COMMENTED** [review]
Looks good, no concerns from my end. Test coverage is good, covers different use patterns with sensible assertions. Verified no required props on the S3Bucket that we should be aware of that might get in the way of using the construct at this level. If something presents, we can address it as part of an issue but all core functionality seems to be easily override-able by the user if needed. Able to discuss any aspects further, but I'm happy with this so far. Nice work 👍

(Let me know if I should approve or if you're just seeking comment-based feedback at this point in time)

**Review by hayesry — DISMISSED** [review]

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 6ef10df07d9d4c3b81c97f602d411f81423e348c
* Result: FAILED
* [Build Logs](https://t2ulkoluaf.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4c6d6eb4-b2b1-4cbb-86fa-9a70813066ca%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 8c0ac63ef357849f77ef543b42ba42fb59bcef6b
* Result: SUCCEEDED
* [Build Logs](https://t2ulkoluaf.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e12d7dea-e99f-4bba-b2b5-27ebc9b6ac96%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Pull request merged**
